warp10/src/main/java/io/warp10/script/functions/DURATIONBUCKETIZE.java

```diff
@@ -0,0 +1,420 @@
+    //
+
+    for (GeoTimeSerie gts : series) {
+      Map<String, String> attributes = gts.getMetadata().getAttributes();
+      if (attributes.get(DURATION_ATTRIBUTE_KEY) != null || attributes.get(OFFSET_ATTRIBUTE_KEY) != null || attributes.get(TIMEZONE_ATTRIBUTE_KEY) != null) {
+        throw new WarpScriptException(getName() + " expects GTS for which the attributes " + DURATION_ATTRIBUTE_KEY + ", " + OFFSET_ATTRIBUTE_KEY + " and " + TIMEZONE_ATTRIBUTE_KEY + " are not set. If an input GTS is supposed to be already duration-bucketized, duration-unbucketize it first before applying a new duration-bucketization.");
+      }
+    }
+
+    //
+    // Check nullity of bucketizer
+    //
+
+    if (null == bucketizer) {
+      throw new WarpScriptException(getName() + " expects a non null bucketizer.");
+    }
+
+    //
+    // Convert duration to joda.time.Period
+    //
+
+    if ('P' != bucketduration.charAt(0)) {
+      throw new WarpScriptException(getName() + " expects that the bucketduration is in ISO8601 duration format.");
+    }
+
+    ADDDURATION.ReadWritablePeriodWithSubSecondOffset bucketperiod;
+    try {
+      bucketperiod = ADDDURATION.durationToPeriod(bucketduration);
+    } catch (WarpScriptException wse) {
+      throw new WarpScriptException(getName() + " encountered an exception: " + wse.getMessage());
+    }
+
+    //
+    // Check that the bucketduration is positive
+    //
+
+    long averageSpan = bucketperiod.getPeriod().toPeriod().toDurationFrom(new Instant()).getMillis() * Constants.TIME_UNITS_PER_MS + bucketperiod.getOffset();
+    if (averageSpan < 0) {
+      throw new WarpScriptException(getName() + " expects the bucketduration parameter to be a positive ISO8601 duration.");
+    }
+
+    //
+    // Compute bucketindex of lastbucket and compute bucketoffset
+    //
+
+    long flag = 0; // always equal to epoch modulo period
+    long bucketoffset;
+    int lastbucketIndex;
+
+    //
+    // Starting from Epoch, we make a hint and land the flag close to lastbucket
+    //
+
+    if (lastbucket > 0) {
+      int lastbucketIndexHint = Math.toIntExact(lastbucket / averageSpan);
+      flag = addNonNegativePeriod(flag, bucketperiod, dtz, lastbucketIndexHint + 1);
+      lastbucketIndex = lastbucketIndexHint;
+
+    } else {
+      int lastbucketIndexHint = - Math.toIntExact(lastbucket / averageSpan);
+      flag = addNonNegativePeriod(flag, bucketperiod, dtz, lastbucketIndexHint);
+      lastbucketIndex = lastbucketIndexHint - 1;
+    }
+
+    //
+    // We move the flag left and right on the time axis to make sure lastbucket is its leftmost bucketend
+    //
+
+    while (flag > lastbucket) {
+      flag = addNonNegativePeriod(flag, bucketperiod, dtz, -1);
+      lastbucketIndex--;
+    }
+
+    while (flag <= lastbucket) {
+      flag = addNonNegativePeriod(flag, bucketperiod, dtz, 1);
+      lastbucketIndex++;
+    }
+
+    bucketoffset = flag - (lastbucket + 1);
```

Comment:
That's working very well to speed-up things for small bucket durations, congrats.
However it seems it is extremely slow with negative last bucket:
```
-1582560100948303 'now' STORE

[
  NEWGTS
  $now NaN NaN NaN T ADDVALUE
  bucketizer.last
  $now 1.2 s +
  'PT1.2S'
  3
]
DURATION.BUCKETIZE
```

Reply:
Bugfix ac3ddee